aten/src/ATen/native/Sorting.cpp

```diff
@@ -183,6 +183,62 @@ std::tuple<Tensor, Tensor> kthvalue(
   return std::make_tuple(values, indices);
 }
 
+Tensor& quantile_out(
+    Tensor& out,
+    const Tensor& self,
+    const Tensor& q,
+    c10::optional<int64_t> _dim,
+    bool keepdim) {
+  
+  TORCH_CHECK(self.numel() > 0, "cannot compute quantile of an empty tensor");
+  TORCH_CHECK(q.dim() <= 1, "q tensor must be a scalar or 1D tensor");
+  TORCH_CHECK(q.numel() > 0, "q tensor is empty");
+
+  TORCH_CHECK(self.device() == q.device(), "input and q tensors must be on the same device");
+  TORCH_CHECK(self.device() == out.device(), "input and output tensors must be on the same device");
+
+  // Using this macro to ensure function is only defined for float dtypes
+  AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "quantile", [&] {
+    TORCH_CHECK(q.options().type_equal(self.options()),
+        "expected q type to be ", self.scalar_type(), " but got ", q.scalar_type());
+    TORCH_CHECK(out.options().type_equal(self.options()),
+        "expected out type to be ", self.scalar_type(), " but got ", out.scalar_type());
+  });
+
+  TORCH_CHECK(q.ge(0).logical_and(q.le(1)).all().item<bool>(),
+      "q values must be in the range [0, 1]");
+
+  auto dim = _dim ? at::maybe_wrap_dim(*_dim, self.dim(), true) : 0;
+  auto sorted = std::get<0>(_dim ? self.sort(dim) : self.flatten().sort());
+
+  auto rank = q * (sorted.size(dim) - 1);
```

Comment:
This will get you in trouble if your tensor is very large, because floating point calculation may not have sufficient resolution to exactly capture precise integer values here. Representable integers in floats get up to `2 ** 24`, beyond that you might end up computing a float that is too big and results in an out of bound access. Generally the best way to solve a problem like this is to split the range if it is too big; another common thing we've done in other cases is just refuse to do the operation if the tensor is too big.

Reply:
For now I'm checking that the size is no bigger than 2 ** 24.